### DISCUSSION TITLE
How do you configure all base path-related settings when serving a Quasar+Vite+Vuejs SPA from nginx behind a proxy?

### DISCUSSION THREAD

**Discussion by matthewadams** [quasarframework/quasar#17037] [General - Components / Directives / etc]
We have an SPA site (@quasar/app-vite@1.4.4) being served from an nginx container that's behind an nginx proxy (in kubernetes), such that any request from the Internet to the proxy for a URL beginning with, say, `https://1.2.3.4/my-ui` returns `http://my-spa.acme.svc.cluster.local:8082` (that dns name is only resolvable inside the cluster).  Therefore, `https://1.2.3.4/my-ui/index.html` causes `http://my-spa.acme.svc.cluster.local:8082/index.html` to be returned, and so on.  We'd like to build our SPA once and only once without knowing what path it will be exposed as via the proxy.  That is `my-ui` may change tomorrow to `the-other-ui`, and we need to be able to configure that at deployment time.

What's the magic configuration of the SPA's base path (and all related configs) so that, when served via the proxy, the site just works?  Recall that the SPA in the browser thinks it's coming from `https://1.2.3.4/my-ui`, but it's really coming from the nginx container behind the proxy.

For reference, our nginx proxy (the one serving `https://1.2.3.4`) has a configuration similar to this:
```
        ...
        location /my-ui/ {
          proxy_pass http://my-spa.acme.svc.cluster.local:8082/;
          proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header X-Forwarded-Proto $scheme;
          proxy_set_header X-Forwarded-Port $server_port;
        }
        ...
```

For the container we serve from inside the cluster (the ones serving `http://my-spa.acme.svc.cluster.local`), we simply copy the SPA's `dist/spa` directory into an nginx container at `/usr/share/nginx/html`, and here is its `nginx.conf`:
```
events{}

http {
  server {
    listen 80;
    http2 on;

    root /usr/share/nginx/html;

    add_header X-Frame-Options "SAMEORIGIN";
    add_header X-XSS-Protection "1; mode=block";
    add_header X-Content-Type-Options "nosniff";

    index index.html;

    charset utf-8;

    location / {
      try_files $uri $uri/ /index.html;
      include  /etc/nginx/mime.types;
    }

    location = /robots.txt  { access_log off; log_not_found off; }

    access_log off;
    error_log  /var/log/nginx/error.log error;

    location ~ /\.(?!well-known).* {
      deny all;
    }
  }
}
```

**Comment by matthewadams** [comment]
The way we configure the runtime aspects of our app is to `await fetch("config.json")`.  `config.json` is a docker volume-mounted file sitting next to `index.html` in our image, and contains, among other things, the base path that we want to prepend to all of our routes.  That way, if a deployer configured the proxy to serve the SPA at `https://1.2.3.4/my-ui`, they'd add `"spaBasePath": "/my-ui"` to the `config.json`, and our [boot file](https://quasar.dev/quasar-cli-webpack/boot-files/) would change the route of every path in the router, prepending the `spaBasePath` to it (basically `it.path = spaBasePath + it.path}` where `it` is each route entry in the router).  Theoretically, are we on the right track?  If so, are there other settings we'd have to change in order for the path modifications to take effect?  We just get a lot of "Path not found" responses no matter what we've tried, and we've tried _a lot_.

**Reply by bbrudastyy** [reply]
Hi, in case the task is still relevant, I was able to solve it as follows.
I have attached the volume in which the env file is stored, which contains the basePath I need. 
When the container starts, a script is run that reads the contents of the file and forms a js file inside the container, inside which variables are imported into the window object.
To read and use the basePath variable, I connected the **_vite-plugin-public-path_** plugin to the collector and configured it so that the js script that is generated at the start of the container does not fall into the rules of this plugin.
Further in **_index.html_** I initially load the script with the environment variables and then set the basePath according to the plugin documentation.
In addition, nginx needs to be slightly configured so that the path to the js script is correct along any route.  I did this by manually proxying
```
    location = /runtime-config.js {
            proxy_pass http://localhost:9000/runtime-config.js;
```

And you also need to slightly modify the routes on the desired location
```
    location /admin-panel/ {
        rewrite ^/admin-panel(/.*)$ $1 break;
        proxy_pass http://localhost:9000;
```

**Reply by matthewadams** [reply]
Yeah....so this just underscores the difficulty of effectively mounting a site at any given context path on a server.  It should not be this hard.
